src/backend/distributed/transaction/multi_transaction.c

```diff
@@ -0,0 +1,1907 @@
+	{
+		connection->sessionLifespan = true;
+	}
+	connection->activeInXact = true;
+
+	return connection;
+}
+
+
+void
+MarkConnectionBusy(MultiConnection *connection)
+{
+	Assert(!connection->busy);
+	connection->busy = true;
+}
+
+
+void
+MarkConnectionIdle(MultiConnection *connection)
+{
+	connection->busy = false;
+}
+
+
+void
+MarkPlacementConnectionFailed(ShardPlacement *placement)
+{
+	ConnectionPlacementHashKey placementKey;
+	ConnectionPlacementHashEntry *placementEntry = NULL;
+	bool found = false;
+
+	/* There seems to be little point in making this work */
+	Assert(InCoordinatedTransaction());
+
+	placementKey.placementid = placement->placementId;
+
+	placementEntry = hash_search(ConnectionPlacementHash, &placementKey, HASH_ENTER, &found);
+	if (!found)
+	{
+		placementEntry->connections = NIL;
+		placementEntry->required = false;
+	}
+
+	placementEntry->failed = true;
+
+	AssociatePlacementShard(placementEntry, placement);
+}
+
+
+void
+MarkPlacementConnectionRequired(ShardPlacement *placement)
+{
+	ConnectionPlacementHashKey placementKey;
+	ConnectionPlacementHashEntry *placementEntry = NULL;
+	bool found = false;
+
+	/* There seems to be little point in making this work */
+	Assert(InCoordinatedTransaction());
+
+	placementKey.placementid = placement->placementId;
+
+	placementEntry = hash_search(ConnectionPlacementHash, &placementKey, HASH_ENTER, &found);
+	if (!found)
+	{
+		placementEntry->connections = NIL;
+		placementEntry->failed = false;
+	}
+
+	placementEntry->required = true;
+
+	AssociatePlacementShard(placementEntry, placement);
+}
+
+
+static void
+AssociatePlacementShard(ConnectionPlacementHashEntry *placementEntry, ShardPlacement *placement)
+{
+	ConnectionShardHashKey shardKey;
+	ConnectionShardHashEntry *shardEntry = NULL;
+	bool found = false;
+	MemoryContext oldContext = NULL;
+
+	shardKey.shardId = placement->shardId;
+	shardEntry = hash_search(ConnectionShardHash, &shardKey, HASH_ENTER, &found);
+	if (!found)
+	{
+		shardEntry->placementConnections = NIL;
+	}
+
+	/* XXX: has to be unique, right? */
+	oldContext = MemoryContextSwitchTo(CacheMemoryContext);
+	shardEntry->placementConnections =
+		list_append_unique_ptr(shardEntry->placementConnections, placementEntry);
+	MemoryContextSwitchTo(oldContext);
+}
+
+
+extern MultiConnection *
+StartPlacementConnection(uint32 flags, ShardPlacement *placement)
+{
+	ConnectionPlacementHashKey key;
+	ConnectionPlacementHashEntry *placementEntry = NULL;
+	MultiConnection *connection;
+	MemoryContext oldContext;
+	bool found;
+
+	key.placementid = placement->placementId;
+
+	/* FIXME: not implemented */
+	Assert(flags & NEW_CONNECTION);
+
+	/*
+	 * Lookup relevant hash entry. We always enter. If only a cached
+	 * connection is desired, and there's none, we'll simply leave the
+	 * connection list empty.
+	 */
+
+	placementEntry = hash_search(ConnectionPlacementHash, &key, HASH_ENTER, &found);
+	if (!found)
+	{
+		placementEntry->connections = NIL;
```

Comment:
On 2016-09-08 07:02:34 -0700, Marco Slot wrote:
> > +
> > +	key.placementid = placement->placementId;
> > +
> > +	/* FIXME: not implemented */
> > +	Assert(flags & NEW_CONNECTION);
> > +
> > +	/*
> > +	 * Lookup relevant hash entry. We always enter. If only a cached
> > +	 * connection is desired, and there's none, we'll simply leave the
> > +	 * connection list empty.
> > +	 */
> > +
> > +	placementEntry = hash_search(ConnectionPlacementHash, &key, HASH_ENTER, &found);
> > +	if (!found)
> > +	{
> > +		placementEntry->connections = NIL;
> 
> When calling GetPlacementConnection from places like the router executor, should we try reuse an existing connection to the same worker here, e.g. default to the first entry ConnectionHash for the worker?

We're doing that, if I understand your idea correctly. We'll use the
underlying host level cache (via StartNodeConnection), to get a possibly
existing connection.   The connections = NIL part is just about
initializing the per-placement connection hash entry.